assets/js/controllers/settings/addresses.controller.js

```diff
@@ -2,75 +2,86 @@ angular
   .module('walletApp')
   .controller('SettingsAddressesCtrl', SettingsAddressesCtrl);
 
-function SettingsAddressesCtrl ($scope, $state, $stateParams, $q, $sce, Wallet, MyWalletHelpers, MyBlockchainApi, Alerts, paymentRequests, $uibModal) {
-  $scope.paymentRequests = paymentRequests;
+function SettingsAddressesCtrl ($scope, $translate, $rootScope, $state, $stateParams, $q, $sce, Wallet, Labels, MyWalletHelpers, MyBlockchainApi, Alerts, $uibModal) {
+  if (!Wallet.status.isLoggedIn) {
+    console.error('Controller depends on being logged in');
+    return;
+  }
+
+  let accountIndex = parseInt($stateParams.account, 10);
+
+  $scope.presentFilter = (address) => address && !!address.label && address.used === false;
+
+  $scope.pastAddressesPage = (page, pageLength) => {
+    let pastFilter = (address) => address && (!address.label || address.used === true);
+
+    return $scope.addresses.filter(pastFilter)
+                           .reverse().slice((page - 1) * pageLength, page * pageLength);
+  };
+
+  $scope.addresses = Labels.all(accountIndex);
+
+  $scope.loading = true;
+
+  Labels.checkIfUsed(accountIndex).then(() => {
+    $scope.loading = false;
+    $rootScope.$safeApply();
+  });
+
   $scope.account = Wallet.accounts()[$stateParams.account];
+
   $scope.receiveIndex = $scope.account.receiveIndex;
 
   $scope.page = 1;
   $scope.pageLength = 20;
-  $scope.totalUsed = $scope.receiveIndex - $scope.paymentRequests.length + 1;
-  $scope.hdLabels = $scope.account.receivingAddressesLabels.reduce((acc, address) => { acc[address.index] = address.label; return acc; }, {});
+  let totalLabeled = $scope.addresses.filter($scope.presentFilter).length;
+  $scope.totalPast = $scope.addresses.length - totalLabeled;
 
   $scope.createAddress = () => {
-    Wallet.addAddressForAccount($scope.account)
-      .then(address => $scope.paymentRequests.push(address))
-      .catch(Alerts.displayError);
+    $scope.createAddressInProgress = true;
+
+    Labels.addLabel(accountIndex, 15, $translate.instant('DEFAULT_NEW_ADDRESS_LABEL'))
+      .then(address => $scope.addresses.push(address))
```

Comment:
I'll double check that, but if I remember correctly, `Labels.all()` returns a clone of the array of addresses, not the original one. So adding an address, wouldn't update the cloned copy. What problem are you seeing?